### ISSUE TITLE
Consider renaming `bin` to `lib`

### ISSUE THREAD

**Issue by RagnarGrootKoerkamp** [RagnarGrootKoerkamp/BAPCtools#136]
Historically `tools.py` was the only and main binary, but by now most files in this directory are just python libraries.
It would be more consistent with other projects to move all python source code to `lib` and make a small bash wrapper script `bin/bt` that can be called from anywhere and invokes `tools.py`.

**Comment by niklasmohrin** [comment]
Maybe it would be even better to follow standard Python layout and put all modules under a common `bapctools` namespace. Going even further, one could make the project pip-installable. In my opinion, having `pip install git+https://github.com/RagnarGrootKoerkamp/BAPCtools.git` (or even `pip install bapctools` if we would publish to PyPI) would be a great improvement of the installation experience. Would you be interested in PRs for that?

**Comment by niklasmohrin** [comment]
> I guess the one possibly tricky thing with external installs is that we have to preserve a lot of non-python files as well, but that's probably not a problem.

Yes, it is possible to include arbitrary other files in a Python wheel and distribute them. I have not used all of the bt subcommands, are there any that require editing the files in the cloned git repository? Because that would correspond to editing files in, say, the Python virtual environment that is used to install bapctools - and editing those files is not usually intended. If the files only need to present for bapctools to read, it's no problem, we would just have to make sure that the code uses relative paths to for example the root file of the bapctools module.

> Does moving everything into a bapctools namespace imply moving everything down one subdirectory?

No, not if we don't want. The layout I would aim for is something like

```
.
├── .git
├── .github
├── .gitignore
├── .pre-commit-config.yaml
├── bapctools
    ├── check_testing_tool.py
    ├── config
    ├── config.py
    ├── conftest.py
    ├── constraints.py
    ├── contest.py
    ├── download_submissions.py
    ├── export.py
    ├── fuzz.py
    ├── generate.py
    ├── headers
    ├── interactive.py
    ├── latex
    ├── latex.py
    ├── misc
    ├── parallel.py
    ├── problem.py
    ├── program.py
    ├── run.py
    ├── skel
    ├── skel.py
    ├── slack.py
    ├── solve_stats.py
    ├── stats.py
    ├── support
    ├── testcase.py
    ├── tools.py
    ├── upgrade.py
    ├── util.py
    ├── validate.py
    ├── validator_tests.py
    ├── verdicts.py
    └── visualize.py
├── bin
├── doc
├── Dockerfile
├── LICENSE
├── pyproject.toml
├── readme.md
├── scripts
├── test
└── third_party
```

The details we can figure out in a PR, but the gist is: Everything that should be distributed is in bapctools/ (except license and readme), because that is how the built wheel distribution will be unpacked into a user's virtual environment. Development helpers, documentation, etc. can stay out, but don't have to; we can select which files to include in the Python package. We can also keep a bin/bt script around that calls the correct Python script so that the current clone-the-repo workflow doesn't break.

Notably, the script is not needed for users installing bapctools with pip if we set up an console script entry point (see first item here: https://packaging.python.org/en/latest/specifications/entry-points/#entry-points).

> Conceptually, I see this as a binary-only package, so I wouldn't necessarily expose the library part of it as standalone thing that can be included by other programs.

Yeah, that's a bit annoying with Python packaging, everything is always exposed. However, I think that if there is no documentation for any submodules of bapctools, it is pretty fair to assume that all of the items in it are implementation details. I don't think people will install bapctools and expect semver correct versioning for some functions somewhere. Even less so if there is no package on PyPI but just an install from Github, which is all I would do in a first step anyways.

There are also plenty of CLI tools that are published on PyPI where the only advertised interface is the CLI, for example https://pypi.org/project/black/

**Comment by niklasmohrin** [comment]
One potential question would be what exactly to name the Python module: bapctools, bapc, bt; it seems to me that bapctools makes the most sense. Even more so when I see that you have already reserved [the bapctools package on PyPI](https://pypi.org/project/bapctools). Although it is possible to publish a module as a package with a different name, it is a bit unconventional and uncalled for in this situation in my opinion. Let me know what you think though, maybe you have other arguments :)

**Comment by niklasmohrin** [comment]
> a big subdir that mixes python and metadata

I understand, I will check what the tooling allows. It is definitely possible to move files around during packaging, for example when using the hatchling build backend: https://hatch.pypa.io/latest/config/build/#rewriting-paths

However, I feel that this would create more anger than joy because the directory layout would be different when running from the repo vs from an installation. There is also [the importlib.resources module](https://docs.python.org/3/library/importlib.resources.html#module-importlib.resources) which could provide a canonical way to do this, but I think it solves a different problem.

The repo-vs-installation discrepancy also applies to ...

> not a fan of `bapctools/bapctools`. Then better `bapctools/lib`

... where we can again tell the tooling that `lib` contains the `bapctools` module, but just running the files would not work without setting them up for this situation. For them, the file config.py would provide the module `lib.config`.

---

I feel that I have missed to describe some of my motivation for this change and what the consequences of it are. In particular, let me try to convince you that `BAPCtools/bapctools` is exactly what is best.

Although the issue here seems to be mostly concerned about the fact that the existing Python files are not really all "bin" files and rather library files, the directory should be renamed. For me, I don't really care about what the directory is called. Instead, I am looking to move the files around so that they adhere to modern Python packaging standards, so that bapctools can benefit from the existing tooling around it. Most importantly, I wouldn't need to give bapctools a special treatment during installation and could instead use pip.

To accomplish this, the repo should provide all the code as part of the bapctools module. All imports in the scripts are also changed to for example `import bapctools.config` (or `from bapctools import config`, so that the following code stays the same). One big reason to have your own code all in one module is so that it does not conflict with other modules from the standard library or other packages. It is also considered a "packaging bug" if a package includes files outside of their module directory (ignoring mechanisms particularly set up for this like importlib.resources).

There are two common layouts to do this: src layout and flat layout (see https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/#src-layout-vs-flat-layout). Since the repo should still be usable without performing an installation, the flat layout is what should be used.

The reason that the _current_ setup works is that when using `python path/to/BAPCtools/bin/tools.py`, the directory of `tools.py` is added to the PYTHONPATH, so all the other files like `config.py` appear as top level modules like `config`. The reason that the flat layout works is that one would run the project from the repository root using for example `python -m bapctools`; when running Python this way, the current working directory is added to the PYTHONPATH and thus the bapctools directory specifies a module. A compatibility script `bin/bt` can be kept around to do exactly that: `cd $0/..; python -m bapctools`. (Note that pip users will have a bt script in their virtual environment's bin directory; they _do not_ have to write `python -m bapctools`)

I understand that it is a bit odd to have the extra bapctools directory when the name of the Git repository is already BAPCtools. However, I don't think that calling it "lib" provides any advantage outside of personal preference. Reusing the name is very common practice, see for example
- https://github.com/psf/black
- https://github.com/pylint-dev/pylint
- https://github.com/django/django

Now, with all this out of the way, my main point is: If files will be moved around, why not go all the way and convert bapctools to a proper Python package? I think that this would open the doors to a cool integration with the Python ecosystem, but I understand if you think that this is not worth it and puts too many constraints on the development. Please let me know what you think :)

**Comment by mpsijm** [comment]
This is awesome, learning a lot here, thank you so much for helping us with this! 🤩 

- Adding a "compatibility script" (I'd rather call it a "convenience script" for the people that do want to run BAPCtools from a cloned repo) in `bin/bt` that does `cd "$(dirname "$0")/.."` followed by `python -m bapctools "$@"` sounds good to me 🙂 
- I'm fine with nesting the non-Python resources that should be part of the package under `bapctools/resources`. Just something we'll have to get used to, but if your editor supports proper quick-search, the length of the path doesn't matter right? 😉 And on the command-line, it's "just" two more quick tab-completes 😛 
- Regarding `checktestdata`: I'd say this binary should be an optional part of the package (most problem packages (also) have a C++ validator), and indeed currently only built for Linux. We can indeed package versions for other systems (see also #459 for work on this), but there will always be system architectures for which we will not provide the `checktestdata` binary. Is it possible to configure our Pip installation in such a way that we have optional architecture-dependent third-party stuff?

**Comment by niklasmohrin** [comment]
> Is it possible to configure our Pip installation in such a way that we have optional architecture-dependent third-party stuff?

As far as I know, no. You distribute Python code in two forms: Source distribution (sdist), or binary distribution (wheel). The former is platform agnostic and used to create the latter. In fact, when "installing" an sdist, pip first builds a wheel from it and then installs that. A wheel on the other hand is basically a zip that contains what should be put in your site-packages, plus some metadata. Usually, sdists don't contain any compiled artifacts, but wheels are meant to contain them. A wheel must also follow a particular naming scheme that includes the target platform, but can also target `any` platform (see https://packaging.python.org/en/latest/specifications/binary-distribution-format/#file-name-convention).

Extras (= optional-dependencies in my PR) are the only thing that can be optional in a Python package, and the only thing they can do is pull in additional dependencies.

So with checktestdata, I would see the following options:
1. Include the compiled artifact in the sdist and forbid building wheels for any platform other than x86 Linux (this should also give an error when trying to install the sdist on other platforms).
2. Package up the compiled checktestdata binary separately and build pure-Python wheel files for bapctools.
3. Somehow build a generic and a Linux specific wheel and only include the binary in the latter, and hope that when installing the correct one is used.

Option 3 seems hacky and is just asking for trouble.

Option 2 is the canonical approach, but it is more effort. An example package that dose it like this is psycopg: They have the main package and also two auxiliary packages psycopg-binary and psycopg-c which can be pulled in via an extra (`pip install psycopg[binary]`, see https://github.com/psycopg/psycopg/blob/b9d533beb5d847ef6837fbd4a011f67730225ffd/psycopg/pyproject.toml#L61). The [main package is pure Python](https://pypi.org/project/psycopg/#files), and [the binary is compiled per platform](https://pypi.org/project/psycopg-binary/#files).

I think that Option 1 is fine for now, unless you plan to provide Python code without the binary to other platforms.

If you ever consider introducing extras, be it for this cause or another, we should consider adding a `full` option or so that always pulls in everything. This way, if we advise users to install bapctools as `bapctools[full]`, their setup won't break if one day you want to change to Option 2, for example. It's also fine to just have everything be a required dependency though.

**Comment by niklasmohrin** [comment]
Licensing could also be interesting. checktestdata is BSD 3-Clause and bapctools is GPL; I guess publishing a separate package would also make this clear :D (I think you should include a BSD license somewhere in the repo anyways?)

**Comment by RagnarGrootKoerkamp** [comment]
I would really prefer if we just add 3 precompiled binaries (windows, mac, Linux) into the bapctools package and select at runtime which one to use. Much less hassle for users that way, and also simpler than making/maintaining a new package for it.

**Comment by mpsijm** [comment]
Fair enough! In either case, that's probably better discussed separately from the rewrite to make BAPCtools Pip-installable, and go with the "pretend that `checktestdata` is any plain resource" strategy for the initial Pip-installable version 🙂

**Comment by mzuenni** [comment]
I think for now it's best to use option 1. but **not** fail on other platforms? (if possible, print a warning that `.ctd` checkers will not work).
- This matches the behavior of the current git
- This allows us to do what ragnar suggested later on without needing to touch the build scripts
